File: rs2-core/src/services/log_reader.rs
//! `log` reader service (PRD §14): query the node log store, scoped to the
//! requesting tenant. Read-only; the URL-params idiom mirrors the `query`
//! service. Logs are operational data — guard the mount with `access` /
//! `manageRoles` like any other.
//!
//! - `GET /<mount>?$take=100&severity=warn&traceId=..&service=/orders&since=..&until=..&q=substr`
//!   → newest-first records (JSON array, or `text/plain` NDJSON via `Accept`).
//! - `GET /<mount>/<traceId>` → all records for one trace (request debugging).

use async_trait::async_trait;
use http::{Method, StatusCode};

use super::{Service, ServiceContext};
use crate::error::RsError;
use crate::logging::{LogQuery, Severity};
use crate::message::{Body, MediaType, Message};

#[derive(Default)]
pub struct LogReaderService;

impl LogReaderService {
    pub fn new() -> Self {
        LogReaderService
    }
}

#[async_trait]
impl Service for LogReaderService {
    async fn handle(&self, msg: Message, ctx: &ServiceContext) -> Result<Message, RsError> {
        if msg.method != Method::GET {
            return Err(RsError::new(
                405,
                crate::error::codes::BAD_REQUEST,
                "Method Not Allowed",
                "the log reader is read-only (GET)",
            ));
        }
        let store = ctx
            .log_store
            .as_ref()
            .ok_or_else(|| RsError::capability_denied("logStore"))?;
        if !store.is_queryable() {
            return Err(RsError::engine_unavailable(
                "logs are exported to an external sink, not locally queryable on this node",
            ));
        }

        let mut q = LogQuery {
            take: msg
                .url
                .query_param("$take")
                .and_then(|v| v.parse::<usize>().ok())
                .unwrap_or(100)
                .min(10_000),
            ..Default::default()
        };
        // Positional `/<traceId>` (request-debugging view); the `traceId`
        // query param overrides it if both are given.
        if let Some(first) = msg.url.service_segments().first() {
            q.trace_id = Some((*first).to_string());
        }
        if let Some(t) = msg.url.query_param("traceId") {
            q.trace_id = Some(t);
        }
        if let Some(s) = msg.url.query_param("severity") {
            match Severity::parse(&s) {
                Some(sev) => q.min_severity = Some(sev),
                None => {
                    return Err(RsError::bad_request(format!(
                        "unknown severity '{s}' (debug|info|warn|error)"
                    )))
                }
            }
        }
        q.service = msg.url.query_param("service");
        q.contains = msg.url.query_param("q");
        q.since = msg.url.query_param("since").and_then(|s| parse_time(&s));
        q.until = msg.url.query_param("until").and_then(|s| parse_time(&s));

        let wants_text = msg
            .header("accept")
            .map(|a| a.contains("text/plain") && !a.contains("application/json"))
            .unwrap_or(false);
        let tenant = msg.tenant.clone();

        let records = store.query(&tenant, &q).await?;
        let total = records.len();

        let body = if wants_text {
            // One OTLP-JSON record per line (NDJSON).
            let text =
                records.iter().map(|r| r.to_otlp_json().to_string()).collect::<Vec<_>>().join("\n");
            Body::from_bytes(text, MediaType::new("text/plain; charset=utf-8"))
        } else {
            let arr = serde_json::Value::Array(records.iter().map(|r| r.to_otlp_json()).collect());
            Body::from_bytes(arr.to_string(), MediaType::json())
        };
        let mut resp = msg.response(StatusCode::OK, Some(body));
        resp.set_header("x-total-count", &total.to_string());
        Ok(resp)
    }
}

/// Parse a `since`/`until` value: Unix milliseconds (integer) or an RFC 3339
/// timestamp. Returns nanoseconds since the epoch to match `timeUnixNano`.
fn parse_time(s: &str) -> Option<u128> {
    if let Ok(ms) = s.parse::<u128>() {
        return Some(ms.saturating_mul(1_000_000));
    }
    time::OffsetDateTime::parse(s, &time::format_description::well_known::Rfc3339)
        .ok()
        .map(|dt| dt.unix_timestamp_nanos().max(0) as u128)
}

File: rs2-worker/src/services/log-reader.ts
// `log` reader service (PRD §14). Port of `rs2-core/src/services/log_reader.rs`.

import { Body } from "../runtime/body";
import { RsError, codes } from "../runtime/error";
import type { LogQuery } from "../runtime/logging";
import { parseSeverity, toOtlpJson } from "../runtime/logging";
import { MediaType } from "../runtime/media-type";
import type { Message } from "../runtime/message";
import type { Service, ServiceContext } from "./context";

/// Parse a `since`/`until` value: Unix milliseconds or RFC 3339, to nanoseconds.
function parseTime(s: string): bigint | undefined {
  if (/^\d+$/.test(s)) return BigInt(s) * 1_000_000n;
  if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$/.test(s)) return undefined;
  const t = Date.parse(s);
  if (Number.isNaN(t)) return undefined;
  return BigInt(Math.max(t, 0)) * 1_000_000n;
}

export class LogReaderService implements Service {
  async handle(msg: Message, ctx: ServiceContext): Promise<Message> {
    if (msg.method !== "GET") {
      throw new RsError(405, codes.BAD_REQUEST, "Method Not Allowed", "the log reader is read-only (GET)");
    }
    const store = ctx.logStore;
    if (!store) throw RsError.capabilityDenied("logStore");
    if (!store.isQueryable()) {
      throw RsError.engineUnavailable("logs are exported to an external sink, not locally queryable on this node");
    }
    const takeParam = msg.url.queryParam("$take");
    const take = Math.min(takeParam !== undefined && /^\+?\d+$/.test(takeParam) ? Number(takeParam) : 100, 10_000);
    const q: LogQuery = { take };
    const first = msg.url.serviceSegments()[0];
    if (first !== undefined) q.traceId = first;
    const traceParam = msg.url.queryParam("traceId");
    if (traceParam !== undefined) q.traceId = traceParam;
    const sev = msg.url.queryParam("severity");
    if (sev !== undefined) {
      const parsed = parseSeverity(sev);
      if (parsed === undefined) throw RsError.badRequest(`unknown severity '${sev}' (debug|info|warn|error)`);
      q.minSeverity = parsed;
    }
    q.service = msg.url.queryParam("service");
    q.contains = msg.url.queryParam("q");
    const since = msg.url.queryParam("since");
    if (since !== undefined) q.since = parseTime(since);
    const until = msg.url.queryParam("until");
    if (until !== undefined) q.until = parseTime(until);

    const accept = msg.header("accept");
    const wantsText = accept !== undefined && accept.includes("text/plain") && !accept.includes("application/json");
    const records = await store.query(msg.tenant, q);
    const total = records.length;
    const body = wantsText
      ? Body.fromString(records.map((r) => JSON.stringify(toOtlpJson(r))).join("\n"), new MediaType("text/plain; charset=utf-8"))
      : Body.fromString(JSON.stringify(records.map(toOtlpJson)), MediaType.json());
    const resp = msg.response(200, body);
    resp.setHeader("x-total-count", String(total));
    return resp;
  }
}
